src/rest_vol.h

```diff
@@ -192,9 +192,198 @@
+    int      num_curlm_msgs    = 0;                                                                          \
+    int      events_occurred   = 0;                                                                          \
+    CURLMsg *curl_multi_msg    = NULL;                                                                       \
+    size_t   fail_count        = 0;                                                                          \
+    size_t   succeed_count     = 0;                                                                          \
+    size_t   num_finished      = 0;                                                                          \
+    size_t   handle_index      = 0;                                                                          \
+                                                                                                             \
+    /*                                                                                                       \
+    Lowers CPU usage dramatically, but also vastly increases time taken for requests to local storage        \
+    when the number of datasets is small.                                                                    \
+    struct timespec delay;                                                                                   \
+    delay.tv_sec  = 0;                                                                                       \
+    delay.tv_nsec = DELAY_BETWEEN_HANDLE_CHECKS;                                                             \
+    */                                                                                                       \
+                                                                                                             \
+    memset(failed_handles_to_retry, 0, sizeof(CURL *) * count);                                              \
+                                                                                                             \
+    do {                                                                                                     \
+        fail_count    = 0;                                                                                   \
+        succeed_count = 0;                                                                                   \
+                                                                                                             \
+        if (CURLM_OK != curl_multi_perform(curl_multi_handle, &num_still_running))                           \
+            FUNC_GOTO_ERROR(H5E_DATASET, H5E_WRITEERROR, FAIL, "cURL multi perform error");                  \
+                                                                                                             \
+        if (CURLM_OK != curl_multi_poll(curl_multi_handle, NULL, 0, 0, &events_occurred))                    \
```

Comment:
Seems to me like we'd probably want to pass an actual timeout value here rather than 0. Docs on this suggest using at least 100ms, but also first checking if [curl_multi_timeout](https://curl.se/libcurl/c/curl_multi_timeout.html) returns a lower value than 100ms and if so, use that instead.

Reply:
Setting a timeout had an issue with always waiting that duration, regardless of activity. Switching to select() instead of poll() let it use the timeout correctly (see new commit).

Reply:
Yes, curl_multi_poll is blocking and so it's going to wait for handle activity or until the timeout has passed. With the new changes, it doesn't look like a call to `select()` was added, unless I missed it. But I think this is achieving the same thing either way, you're still going to wait on the handle for the given timeout that you give to the `select()` call. I think using `curl_multi_poll()` is just cleaner here. A decision needs to be made as to whether there is useful work to do if a 0 timeout is going to be used. Otherwise I think you want to use a small timeout like 100ms regardless to give curl time to process requests and not keeping making `curl_multi_poll` calls that return immediately after doing nothing.

Reply:
Looks like I accidentally deleted the select line before committing; it should be there now.

I tested the time `curl_multi_poll()` and the overall vol tests took with a few timeout values. With 0 ms the vol tests took about 12 seconds, while a 10 ms default timeout (lowered if _timeout returned a lower value) took it up to ~36 seconds. Checking each iteration showed that `curl_multi_poll()` was always taking at least 1 ms to wait. The documentation for poll says it should return before the timeout if activity is detected, so I suspect that something more needed to be setup for it to detect activity on the handles.

The new select() method waits for the timeout on some iterations, but returns near instantly on others, so the timeout seems to be returning early when activity is detected now.